apps/prairielearn/src/lib/groups.ts

```diff
@@ -196,8 +196,19 @@ export async function joinGroup(
     flash('error', 'The join code has an incorrect format');
     return;
   }
+
   const groupName = splitJoinCode[0];
   const joinCode = splitJoinCode[1].toUpperCase();
+
+  // This is a best-effort check to produce a nice error message. Even if this
+  // fails due to a race condition, the `group_users_insert` sproc below will
+  // validate that the user isn't already in a group.
```

Comment:
I used this pattern throughout this PR: optimistically try to detect multiple memberships, but still pessimistically enforce uniqueness via locking in sprocs.